### PR TITLE
Jv spring boot project hw 3

### PR DESCRIPTION
**Pull request by ihorvasylenko23**

### CODE REVIEW

**Review by SaiyaJ1N — APPROVED** [review]
LGTM!!

File: src/main/java/mate/academy/repositiry/book/BookSpecificationBuilder.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+@Component
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    @Autowired
+    private SpecificationProviderManager<Book> bookSpecificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParametersDto searchParametrers) {
+        Specification<Book> spec = Specification.where(null);
+        if (searchParametrers.authors() != null && searchParametrers.authors().length > 0) {
```
Comment (vovasalyha) [inline src/main/java/mate/academy/repositiry/book/BookSpecificationBuilder.java]: you are duplicating the same code here many times. for me it's a good candidate for private method.

File: src/main/java/mate/academy/repositiry/book/BookSpecificationProviderManager.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+@Component
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+    private List<SpecificationProvider<Book>> bookSpecificationProviders;
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        return bookSpecificationProviders.stream()
+                .filter(p -> p.getKey().equals(key))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException("Can`t find correct "
```
Comment (vovasalyha) [inline src/main/java/mate/academy/repositiry/book/BookSpecificationProviderManager.java]: better use `NoSuchElementException`

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                .orElseThrow(() -> new EntityNotFoundException("Can`t find employee by id: " + id));
+        return bookMapper.toDto(book);
+    }
+
+    @Override
+    public BookDto update(Long id, CreateBookRequestDto requestDto) {
+        Book existingBook = bookRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find book by id: " + id));
+
+        Book updatedBook = bookMapper.toModel(requestDto);
```
Comment (vovasalyha) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java]: if you are fetching existing book, just set new fields for that book entity which you fetched, you don't need to create new object.

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        updatedBook.setId(existingBook.getId());
+
+        return bookMapper.toDto(bookRepository.save(updatedBook));
+    }
+
+    @Override
+    public void delete(Long id) {
+        if (!bookRepository.existsById(id)) {
```
Comment (vovasalyha) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java:66]: in most cases it's ok if entity is not found, just nothing will happen, but it's for you to decide.

**Review by vovasalyha — CHANGES_REQUESTED** [review]
GJ. Take a look on few suggestions. Also, you have here changes from all your PRs. Please check Bohdan's guide about several PRs at the time please.

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+databaseChangeLog:
+  - include: 
+      file: db/changelog/changes/01-create-books-table.yaml
```
Comment (Vasul-Pankiv) [inline src/main/resources/db/changelog/db.changelog-master.yaml:3]: add empty line to the end of file

**Review by Vasul-Pankiv — APPROVED** [review]

**Pull request merged**